[10.0][MIG] stock_inventory_lockdown

MiquelRForgeFlow opened the pull request:
Standard migration.

- [x] Added https://github.com/OCA/stock-logistics-warehouse/pull/304.

Commit:
New module stock_inventory_lockdown

This module is a feature extracted from stock_inventory_location.

Ported to v8 and new API
Adopted latest OCA standards
Added tests

oca-transbot committed:
OCA Transbot updated translations from Transifex

MiquelRForgeFlow committed:
stock_inventory_lockdown

oca-transbot committed:
OCA Transbot updated translations from Transifex

LoisRForgeFlow reviewed (commented):
I just tested on runbot and realized of one issue, which is that the lockdown is not taking into account the children of the location where an inventory adjustment is being conducted. For instance:

1. Start an inventory Adjustment (IA) in Stock.
2. Try to deliver some apple headphones (which are stored in Stock/Shelf1).
3. No warning raises and the picking is completed.

At the moment this module works perfect if you are using `stock_inventory_exclude_sublocation` and excluding sublocations in IA. But not with standard odoo.

JordiBForgeFlow committed:
refactor the module to perform lockdown validations using constraint
 in stock move instead of checks in quants. This will allow for a more
 robust lockdown, and still make it possible to perform inventory adjustments
 in the locked location. Also resolves an outstanding issue related to the
 previous design not allowing inventory adjustments where negative quants existed.

LoisRForgeFlow commented:
Pending to solve this: https://github.com/OCA/stock-logistics-warehouse/pull/304

MiquelRForgeFlow committed:
[MIG] stock_inventory_lockdown: Migration to 10.0

MiquelRForgeFlow commented:
https://github.com/OCA/stock-logistics-warehouse/pull/304 was merged, so this PR is ready to be reviewed and merged.

serpentcs-dev1 commented on stock_inventory_lockdown/models/stock_location.py:
```diff
@@ -0,0 +1,45 @@
+    @api.model
+    def create(self, vals):
+        """Refuse create if an inventory is being conducted at the parent"""
+        if 'location_id' in vals:
```
This is giving error when you try to create a new stock location and when there is an 'In Progress' Inventory, because here the if the condition gets satisfied even when `location_id` is empty.

This should be like `if 'location_id' in vals and vals['location_id']:` or simply `if vals.get('location_id'):` like you have done in write method.

serpentcs-dev1 commented on stock_inventory_lockdown/models/stock_inventory.py:
```diff
@@ -0,0 +1,31 @@
+            ('location_id', 'child_of', location_ids.ids),
+            ('usage', 'in', ['internal', 'transit'])]
+        if locations_ids:
+            location_domain.append(('location_id', 'child_of', locations_ids))
```
The domain  `('location_id', 'child_of', locations_ids)` is already added above in the list!!

MiquelRForgeFlow replied:
<code>location_ids</code> != <code>locations_ids</code>

serpentcs-dev1 replied:
@mreficent I am talking about the [line 27](https://github.com/OCA/stock-logistics-warehouse/pull/299/files/4f391ba257942d07fb874f32b16d2483483bc339#diff-3817be29f13c1c2ff56a0a821a025338R27) the child locations are already appended to the domain list.

MiquelRForgeFlow replied:
@darshan-serpent

in line 18: you have `locations_ids`, added in inventory_domain
in line 27: you have `location_ids.ids`
in line 30: you have `locations_ids`, added in location_domain

The three things are different.

serpentcs-dev1 replied:
Oh Yes right, now I noticed, thanks! :smile:

nikul-serpentcs commented on stock_inventory_lockdown/README.rst:
```diff
@@ -0,0 +1,60 @@
+=====
+
+.. image:: stock_inventory_lockdown/static/images/move_error.png
+   :alt: Error message
```
@mreficent This link does not work, Can you please take a look?

MiquelRForgeFlow replied:
There is no way to solve this. We can put it in a way that it works in github or in a way it works in Odoo, but not in both.

rousseldenis commented on stock_inventory_lockdown/models/stock_location.py:
```diff
@@ -0,0 +1,45 @@
+                  'location'))
+
+    @api.multi
+    def write(self, vals):
```
@mreficent I know this is migration, but shouldn't be checked through constrains ? Moreover that avoid write and create overrides.

MiquelRForgeFlow replied:
@rousseldenis  Will it work also for the unlink method?

rousseldenis replied:
Of course not, but you can call the _check_inventory in a constrains method and call the _check in unlink specifically

MiquelRForgeFlow commented:
Please, could everyone update their reviews?

LoisRForgeFlow commented on stock_inventory_lockdown/models/stock_location.py:
```diff
@@ -0,0 +1,32 @@
+    @api.multi
+    def unlink(self):
+        """Refuse unlink if an inventory is being conducted"""
+        location_inventory_open_ids = self.env['stock.inventory'].sudo(). \
```
This is not working as expected, maybe due to the space after `sudo()`, please remove it.

LoisRForgeFlow commented on stock_inventory_lockdown/models/stock_location.py:
```diff
@@ -0,0 +1,32 @@
+    _order = 'name'
+
+    @api.multi
+    @api.constrains('location_id')
```
This is also not working as it worked before, raises on creation but not on write...

rousseldenis replied:
Weird as the trigger is location_id.

The code is different from the initial migration commit.

Here you don't take into account self in recordset to check.

MiquelRForgeFlow replied:
But if we change the `_check_inventory_location_id` to check the self, then the create won't work correctly :S

rousseldenis commented:
@mreficent Can you add test on the unlink to get good coverage ?

LoisRForgeFlow reviewed (approved):
This is working fine now! :+1:

LoisRForgeFlow commented:
@nikul-serpentcs @darshan-serpent could you update your reviews please?

serpentcs-dev1 reviewed (approved):
Code Review LGTM :+1:

nikul-serpentcs reviewed (approved):
Functional Testing, Works Fine!!! :+1:
